test/IdRegistry/IdRegistry.st.sol

```diff
@@ -0,0 +1,192 @@
+
+import {IdRegistryHarness} from "../Utils.sol";
+
+contract IdRegistrySymTest is SymTest, Test {
+    IdRegistryHarness idRegistry;
+    address trustedCaller;
+
+    address x;
+    address y;
+
+    function setUp() public {
+        // Setup IdRegistry
+        idRegistry = new IdRegistryHarness(address(0));
+
+        trustedCaller = address(0x1000);
+        idRegistry.setTrustedCaller(trustedCaller);
+
+        // Register fids
+        vm.prank(trustedCaller);
+        idRegistry.trustedRegister(address(0x1001), address(0x2001));
+        vm.prank(trustedCaller);
+        idRegistry.trustedRegister(address(0x1002), address(0x2002));
+
+        assert(idRegistry.idOf(address(0x1001)) == 1);
+        assert(idRegistry.idOf(address(0x1002)) == 2);
+
+        assert(idRegistry.getRecoveryOf(1) == address(0x2001));
+        assert(idRegistry.getRecoveryOf(2) == address(0x2002));
+
+        // Create symbolic addresses
+        x = svm.createAddress("x");
+        y = svm.createAddress("y");
+    }
+
+    // Additional setup to cover various input states
+    function init() public {
+        if (svm.createBool("disableTrustedOnly?")) {
+            idRegistry.disableTrustedOnly();
+        }
+        if (svm.createBool("pauseRegistration?")) {
+            idRegistry.pauseRegistration();
+        }
+    }
+
+    // Verify the IdRegistry invariants
+    function check_invariant(bytes4 selector, address caller) public {
+        init();
+
+        // Consider two distinct addresses
+        vm.assume(x != y);
+
+        // Record pre-state
+        uint256 oldIdX = idRegistry.idOf(x);
+        uint256 oldIdY = idRegistry.idOf(y);
+
+        uint256 oldIdCounter = idRegistry.getIdCounter();
+
+        bool oldPaused = idRegistry.paused();
+
+        // Execute an arbitrary tx to IdRegistry
+        vm.prank(caller);
+        (bool success,) = address(idRegistry).call(mk_calldata(selector));
+        vm.assume(success); // ignore reverting cases
+
+        // Record post-state
+        uint256 newIdX = idRegistry.idOf(x);
+        uint256 newIdY = idRegistry.idOf(y);
+
+        uint256 newIdCounter = idRegistry.getIdCounter();
+
+        // Verify invariant properties
+
+        // Ensure that there is no recovery address associated with fid 0.
+        assert(idRegistry.getRecoveryOf(0) == address(0));
+
+        // Ensure that idCounter never decreases.
+        assert(newIdCounter >= oldIdCounter);
+
+        // If a new fid is registered, ensure that:
+        // - IdRegistry must not be paused.
+        // - idCounter must increase by 1.
+        // - The new fid must not be registered for an existing fid owner.
+        // - The existing fids must be preserved.
+        if (newIdCounter > oldIdCounter) {
+            assert(oldPaused == false);
+            assert(newIdCounter - oldIdCounter == 1);
+            assert(newIdX == oldIdX || oldIdX == 0);
+            assert(newIdX == oldIdX || newIdY == oldIdY);
```

Comment:
I found these last two lines hard to reason about, and rewrote them as follows. From the POV of symbolic verification, is there any difference between what we are asserting here?

```
            if (newIdX != oldIdX) {
                assert(oldIdX == 0 && newIdX == newIdCounter);
                assert(oldIdY == newIdY);
            } else if (newIdY != oldIdY) {
                assert(oldIdY == 0 && newIdY == newIdCounter);
                assert(oldIdX == newIdX);
            } else {
                assert(oldIdX == newIdX && oldIdY == newIdY);
            }
```

Reply:
very helpful thanks.

1. Agree that the last branch is unnecessary, and can be removed. 
2. What are the implications of logical duality here? Does it imply that this code is testing different things than the original logic you specified?